File: src/solutions/java/40.combination-sum-ii.java
package solutions.java;

import java.util.*;
/*
 * @lc app=leetcode id=40 lang=java
 *
 * [40] Combination Sum II
 */

// @lc code=start
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> ans = new LinkedList<>();
        List<Integer> curr = new LinkedList<>();
        Arrays.sort(candidates);
        dfs(0, target, curr, ans, candidates, -1);
        return ans;
    }

    public void dfs(int sum, int target, List<Integer> curr, List<List<Integer>> ans, int[] candidates, int start) {
        for (int i = start + 1; i < candidates.length; i++) {
            if (sum + candidates[i] > target || (i >= start + 2 && candidates[i] == candidates[i - 1])) {
                continue;
            }
            sum += candidates[i];
            curr.add(candidates[i]);
            if (sum == target) {
                ans.add(new LinkedList<>(curr));
            } else if (sum < target) {
                dfs(sum, target, curr, ans, candidates, i);
            }
            sum -= candidates[i];
            curr.remove(curr.size() - 1);
        }
    }
}
// @lc code=end

File: src/cpp/40.combination-sum-ii.cpp
/*
 * @lc app=leetcode id=40 lang=cpp
 *
 * [40] Combination Sum II
 *
 * https://leetcode.com/problems/combination-sum-ii/description/
 *
 * algorithms
 * Medium (52.52%)
 * Likes:    5710
 * Dislikes: 147
 * Total Accepted:    579.1K
 * Total Submissions: 1.1M
 * Testcase Example:  '[10,1,2,7,6,1,5]\n8'
 *
 * Given a collection of candidate numbers (candidates) and a target number
 * (target), find all unique combinations in candidates where the candidate
 * numbers sum to target.
 *
 * Each number in candidates may only be used once in the combination.
 *
 * Note: The solution set must not contain duplicate combinations.
 *
 *
 * Example 1:
 *
 *
 * Input: candidates = [10,1,2,7,6,1,5], target = 8
 * Output:
 * [
 * [1,1,6],
 * [1,2,5],
 * [1,7],
 * [2,6]
 * ]
 *
 *
 * Example 2:
 *
 *
 * Input: candidates = [2,5,2,1,2], target = 5
 * Output:
 * [
 * [1,2,2],
 * [5]
 * ]
 *
 *
 *
 * Constraints:
 *
 *
 * 1 <= candidates.length <= 100
 * 1 <= candidates[i] <= 50
 * 1 <= target <= 30
 *
 *
 */

// @lc code=start
class Solution {
public:
  vector<vector<int>> combinationSum2(vector<int> &candidates, int target) {
    vector<vector<int>> ans;
    if (candidates.size() == 0) {
      return ans;
    }
    sort(candidates.begin(), candidates.end());
    vector<int> curr;
    solve(candidates, ans, curr, target, 0, 0);
    return ans;
  }
  void solve(vector<int> &candidates, vector<vector<int>> &ans,
             vector<int> &curr, int target, int sum, int index) {
    if (sum == target) {
      ans.emplace_back(curr);
      return;
    }
    if (sum > target) {
      return;
    }
    for (int i = index; i < candidates.size(); i++) {
      if (i > index && candidates[i] == candidates[i - 1]) {
        continue;
      }
      curr.emplace_back(candidates[i]);
      solve(candidates, ans, curr, target, sum + candidates[i], i + 1);
      curr.pop_back();
    }
  }
};
// @lc code=end
